Examples/MemoizedWebCrawler/README.md

```diff
@@ -0,0 +1,310 @@
+            group: group,
+            db: db,
+            functionCache: functionCache,
+            executor: executor
+        )
+        let results = try await engine.build(key: FetchTitle(url: "http://example.com/"), ctx).get()
+        XCTAssertEqual(results.pageTitle, "Example Domain")
+    }
+}
+```
+
+#### **🚧 Example 2: Fetch image URLS + Fetch image blobs**
+
+Your solo business is doing great and you want to ship a new functionality. You want to extend your crawler to extract the list of all images from the HTML. At this point you can disconnect from the internet and rely on the fact that `FetchHTTP` is already memoized. If you want to implement `FetchListOfImages` on the airplane, you have the option to do so. llbuild2fx has populated the CAS with the right objects. On top of that it has populated the function cache, which maps already-computed keys to objects in CAS.
+
+Here is how you would implement `FetchListOfImages` and write a test for it:
+
+```
+TBD
+```
+
+### Decoupling “what” from “how”: A top-level “Build” key
+
+```swift
+public struct BuildResult: Codable {
+    let result: TypedBuildResult
+}
+
+extension BuildResult: FXValue {}
+
+public enum TypedBuildResult: Codable {
+    case txtFile(DataID)
+    case tarFile(DataID)
+}
+
+public struct Build: AsyncFXKey, Encodable {
+    public typealias ValueType = BuildResult
+
+    public static let version: Int = 1
+    public static let versionDependencies: [FXVersioning.Type] = [Build.self]
+
+    let goal: String
+
+    public func computeValue(_ fi: FXFunctionInterface<Self>, _ ctx: Context) async throws -> BuildResult {
+        if goal == "release.txt" {
+            // Fetch release.txt via HTTP
+        } else if goal == "release.tar" {
+            // ...
+        } else if goal == "foo.html" {
+            // ...
+        } else if goal == "bar.png" {
+            // ...
+        } else if goal.hasPrefix("http://") {
+            // ...
+        }
+    }
+}
+```
+
+### 🚧 Example 3:  Shake’s release.tar example
+
+Another illuminating example is Shake's `release.tar` example. Consider either of these scenarios:
+
+* suppose you are implementing a static site generator, and as part of a final step, you want to create a tar file and scp+untar it to a remote server.
+* suppose you maintain an open source project and  you want to release sources for the most recent version as a tar file on web. Two important observations:
+    * the list of files to be included in the final compressed file is listed in `release.txt` (which can itself be dynamically generated)
+    * a file listed in `release.txt` may not be a leaf file and needs to be “built” on the fly.
+
+Here’s how you would express this in Shake:
+
+```haskell
+import Development.Shake
+import Development.Shake.FilePath
+
+main = shakeArgs shakeOptions $ do
+    want ["result.tar"] -- (1)
+    "*.tar" %> \out -> do -- (2)
+        contents <- readFileLines $ out -<.> "txt" -- (3)
```

Comment:
> `readFile' :: Partial => FilePath -> Action String`
> Read a file, after calling [need](https://hackage.haskell.org/package/shake-0.19.8/docs/Development-Shake.html#v:need). The argument file will be tracked as a dependency.

and

> `readFileLines :: Partial => FilePath -> Action [String]`
> A version of [readFile'](https://hackage.haskell.org/package/shake-0.19.8/docs/Development-Shake.html#v:readFile-39-) which also splits the result into lines. The argument file will be tracked as a dependency

From https://hackage.haskell.org/package/shake-0.19.8/docs/Development-Shake.html#v:readFile-39-